src/librustc_mir/transform/check_consts/validation.rs

```diff
@@ -33,42 +33,72 @@ type IndirectlyMutableResults<'mir, 'tcx> =
 type QualifResults<'mir, 'tcx, Q> =
     dataflow::ResultsCursor<'mir, 'tcx, FlowSensitiveAnalysis<'mir, 'mir, 'tcx, Q>>;
 
+#[derive(Default)]
 pub struct Qualifs<'mir, 'tcx> {
-    has_mut_interior: QualifResults<'mir, 'tcx, HasMutInterior>,
-    needs_drop: QualifResults<'mir, 'tcx, NeedsDrop>,
-    indirectly_mutable: IndirectlyMutableResults<'mir, 'tcx>,
+    has_mut_interior: Option<QualifResults<'mir, 'tcx, HasMutInterior>>,
+    needs_drop: Option<QualifResults<'mir, 'tcx, NeedsDrop>>,
+    indirectly_mutable: Option<IndirectlyMutableResults<'mir, 'tcx>>,
 }
 
 impl Qualifs<'mir, 'tcx> {
     fn indirectly_mutable(
         &mut self,
-        _: &Item<'mir, 'tcx>,
+        item: &'mir Item<'mir, 'tcx>,
         local: Local,
         location: Location,
     ) -> bool {
-        self.indirectly_mutable.seek_before(location);
-        self.indirectly_mutable.get().contains(local)
+        let indirectly_mutable = self.indirectly_mutable.get_or_insert_with(|| {
+            let Item { tcx, body, def_id, param_env, .. } = *item;
+
+            // We can use `unsound_ignore_borrow_on_drop` here because custom drop impls are not
+            // allowed in a const.
+            //
+            // FIXME(ecstaticmorse): Someday we want to allow custom drop impls. How do we do this
```

Comment:
we can lower `Drop` terminators for types that have a `const Drop` bound to a function call before doing this analysis.